ext/mbstring/mbstring.c

```diff
@@ -4800,11 +4809,355 @@ finish_up_remaining_bytes: ;
+		_1BYTE | OVERLONG_3BYTE | INVALID_CP, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | SURROGATE | INVALID_CP, _1BYTE | SURROGATE | INVALID_CP,
+		_1BYTE | SURROGATE | INVALID_CP, _1BYTE | SURROGATE | INVALID_CP,
+		_1BYTE | SURROGATE | INVALID_CP, _1BYTE | SURROGATE | INVALID_CP,
+		_1BYTE | OVERLONG_3BYTE | INVALID_CP, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		_1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE, _1BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE);
+
+	const __m256i find_continuation = _mm256_set1_epi8(-64);
+	const __m256i _b = _mm256_set1_epi8(0xB);
+	const __m256i _d = _mm256_set1_epi8(0xD);
+	const __m256i _f = _mm256_set1_epi8(0xF);
+
+	__m256i last_hi_nibbles = _mm256_setzero_si256(), last_lo_nibbles = _mm256_setzero_si256();
+	__m256i operand;
+
+	while (p < e) {
+		operand = _mm256_loadu_si256((__m256i*)p);
+
+check_operand:
+		if (!_mm256_movemask_epi8(operand)) {
+			/* Entire 32-byte block is ASCII characters; the only thing we need to validate is that
+			 * the previous block didn't end with an incomplete multi-byte character
+			 * (This will also confirm that the previous block didn't end with a bad byte like 0xFF) */
+			__m256i bad_mask = _mm256_set_epi8(0xB, 0xD, 0xE, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127);
+			__m256i bad = _mm256_cmpgt_epi8(last_hi_nibbles, bad_mask);
+			if (_mm256_movemask_epi8(bad)) {
+				return false;
+			}
+
+			/* Consume as many full blocks of single-byte characters as we can */
+			while (true) {
+				p += sizeof(__m256i);
+				if (p >= e) {
+					goto finish_up_remaining_bytes;
+				}
+				operand = _mm256_loadu_si256((__m256i*)p);
+				if (_mm256_movemask_epi8(operand)) {
+					break;
+				}
+			}
+		}
+
+		__m256i hi_nibbles = _mm256_and_si256(_mm256_srli_epi16(operand, 4), _f);
+		__m256i lo_nibbles = _mm256_and_si256(operand, _f);
+
+		__m256i lo_nibbles2 = _mm256_shift_epi8(last_lo_nibbles, lo_nibbles, 1);
+		__m256i hi_nibbles2 = _mm256_shift_epi8(last_hi_nibbles, hi_nibbles, 1);
+
+		__m256i bad = _mm256_cmpgt_epi8(_mm256_and_si256(_mm256_and_si256(_mm256_shuffle_epi8(bad_lo_nibble2, lo_nibbles2), _mm256_shuffle_epi8(bad_hi_nibble2, hi_nibbles2)), _mm256_shuffle_epi8(bad_hi_nibble, hi_nibbles)), _mm256_setzero_si256());
```

Comment:
Maybe add some line breaks to this line as it's *dense*

Reply:
Let me do that right now. Thanks.

Reply:
Haven't pushed an update yet... will do so soon. But line breaks have been added in my local copy.